### ISSUE TITLE
Rule proposal: Prefer computed properties

### ISSUE THREAD

**Issue by SimenB** [eslint/eslint#7377]
**Please describe what the rule should do:**
It would warn when assignment to properties whose names are variables could instead use ES6 computed properties.

Airbnb style guide recommends this (https://github.com/airbnb/javascript#objects, 3.2)

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

``` js
function tupleToObject(arg) {
  const o = {};

  o[arg[0]] = arg[1];

  return o;
}
```

Correct code:

``` js
function tupleToObject(arg) {
  return { [arg[0]]: arg[1] };
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Eslint already warns about unnecessary computed properties (`no-useless-computed-key`), it should also report when it's "necessary", or should be preferred.

Off topic: why is there a naming mismatch between  `no-useless-computed-key` and a `computed-property-spacing`? `key` or `property`? I realize changing it would mean a breaking change, but they should be consistent (I prefer `property`, FWIW).

I tried searching for this (I find it odd no one has requested this before...), but was unable to find an issue for it.

**Comment by platinumazure** [comment]
This would require some type analysis, which ESLint is not really capable of.

Let me illustrate:

```js
let o = {};

// Some code

o = getValue();

o[foo] = bar; // I assume this shouldn't warn
```

And it would get even better when closures are involved.

So I think either the proposal needs to be fleshed out a little more to reflect exactly when the rule should or shouldn't warn.

And at this point, if my assumption (that the rule can only warn if the most recent object assignment was an object literal, before the computed property assignment) is true, I would :-1: the proposal because I don't think it would provide enough value-add to users to be worth the maintenance cost of adding a 230-something-eth rule. (We ask for new rule submissions: Can ESLint be said to be incomplete without this rule? I'm personally not seeing a strong case here.)

**Comment by SimenB** [comment]
Since it would only warn at the point of instantiation, it wouldn't be an issue for eslint to figure out `o` has been reassigned, would it?

It only makes sense if `o` is a `const ` (or could be `const`).

**Comment by nzakas** [comment]
I'm also 👎 . I don't think the use case is generic enough for a core rule, but you can certainly always make a custom rule if this is important to you.

**Comment by not-an-aardvark** [comment]
It looks like this proposal hasn't had any activity in awhile, and it has two :-1:s from team members, so I'm going to go ahead and close this issue. Of course, you can always use this as a custom rule.
